### PR TITLE
eyre: add channel \n heartbeat every 20 seconds

### PR DESCRIPTION
**Pull request by pkova**

### CODE REVIEW

**Comment by ohAitch** [comment]
Ooh hadn't caught that the eyre rewrite finally uses Server Sent Events,
nice!

**Comment by loganallenc** [comment]
This looks good to me. What do you think @eglaysher?

File: pkg/arvo/sys/vane/eyre.hoon [hunk 1]
```diff
@@ -1194,6 +1193,12 @@
       ^-  move
       [duct %pass /channel/timeout/[channel-id] %b %wait expiration-time]
     ::
+    ++  heartbeat-move
+      |=  [channel-id=@t]
+      ^-  move
+      :^  duct  %pass  /channel/heartbeat/[channel-id]
+      [%b %wait (add now ~s20)]
```
Comment (eglaysher) [inline pkg/arvo/sys/vane/eyre.hoon]: You'll probably want a matching `+cancel-heartbeat-move` for when a channel closes so there isn't a queued heartbeat. Yes, you check for the existence of the channel in `+on-channel-heartbeat`, but better to never receive the wakeup in the first place.

File: pkg/arvo/sys/vane/eyre.hoon [hunk 2]
```diff
@@ -2103,6 +2117,23 @@
       =^  moves  server-state.ax
         (on-channel-timeout i.t.t.wire)
       [moves http-server-gate]
+    ::
+        %heartbeat
+      =/  on-channel-heartbeat
+        on-channel-heartbeat:by-channel:(per-server-event event-args)
+      =/  handle-response
```
Comment (eglaysher) [inline pkg/arvo/sys/vane/eyre.hoon]: You probably want to move most of this logic into `+on-channel-heartbeat` itself. You have access to `+handle-response` there and the logic handling this probably shouldn't be split into two places.

**Comment by jtobin** [comment]
@pkova Did you want to update this one?  It's been sitting here for almost a month, so I'll close it soon unless you reckon you'll take it further.

**Comment by jtobin** [comment]
Alternatively, this might be worth merging as-is anyway, as it's probably a strict improvement.

(cc @eglaysher)

**Comment by pkova** [comment]
I'll update this during the upcoming week.

Haven't gotten around to it due to being busy and having to learn some new stuff to implement the proposed changes (mainly `++load` for new heartbeat state).

Hopefully this isn't blocking anything.

**Comment by jtobin** [comment]
No sweat, I was just gonna clean it up if it was going nowhere.  Take all the time you need.

**Comment by pkova** [comment]
Alright, implemented the proposed changes now. I'm having trouble creating a pill or even running `make test` locally, so I ask someone with a more blessed computer to do that for me. Got the same error message as [here.](https://github.com/urbit/urbit/pull/1777#issuecomment-537326130)

**Comment by yosoyubik** [comment]
@pkova I've been using this https://github.com/urbit/urbit/issues/1779#issuecomment-538646101, suggested byt @jtobin and it has worked for me:

```
$ urbit -F zod -B bin/solid.pill -A pkg/arvo
[..]
~zod:dojo> .solid/pill +solid
[..]
$ cp zod/.urb/put/solid.pill bin
```

**Comment by jtobin** [comment]
@pkova I've PR'd a pill your way.  The tests pass, so as soon as @eglaysher takes another look I'll merge this in.

(I'll look into the weird business going on with Nix and pills as of late, too.)

**Pull request merged**